### PR TITLE
fix: Pass block explicitly in `define_method` calls for PG instrumentation query methods

### PR DESCRIPTION
**Pull request by Capncavedan**
https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/570

"As a result of the fact blocks are captured by closures they cannot be passed implicitly to methods created with `define_method`; they must instead be passed explicitly using the `define_method(meth) { |args, &block| ... }` syntax."
from https://banisterfiend.wordpress.com/2010/11/06/behavior-of-yield-in-define_method/

This PR changes the "exec-ish" and "exec-prepared-ish" methods to all optionally handle an explicit block, with a new separate test file specifically for these PG::Connection patches.

This PR does not change the "prepare-ish" methods as those don't seem to return anything anyway.

### CODE REVIEW

**Comment by arielvalentin** [comment]
@Capncavedan thanks again for your help!

Please review and sign the Easy CLA so that we may start the PR review process.

**Comment by Capncavedan** [comment]
@arielvalentin done

**Comment by arielvalentin** [comment]
Thank you. Did you happen to have the ability to test this in a sample app? 

If so would you be able to share some screen shots of the fix in action?

**Comment by Capncavedan** [comment]
@arielvalentin Only in the repro script so far.  This bug manifested up in a background process of ours, and we patched it to process its Postgres queries differently.

Apologies if this is already clear - the bug isn't affecting telemetry, rather one specific way of issuing Postgres queries using the PG gem.

File: instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'test_helper'
+require 'active_record'
```
Comment (arielvalentin) [inline instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb:8]: Where is `active_record` being used in these tests?

File: instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+require 'pg'
+
+require_relative '../../../../../lib/opentelemetry/instrumentation/pg'
+require_relative '../../../../../lib/opentelemetry/instrumentation/pg/patches/connection'
```
Comment (arielvalentin) [inline instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb:12]: Are these requires necessary? i.e. is the `test_helper` and bundler not requiring these already?

File: instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+describe OpenTelemetry::Instrumentation::PG::Patches do
```
Comment (arielvalentin) [inline instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb:14]: ✨ Thank you for adding tests for this!

**Review by arielvalentin — COMMENTED** [review]
Thank you for this contribution!

Please update your commit messages to use conventional commits: https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/CONTRIBUTING.md#use-conventional-commit-messages

File: instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb (hunk 2 above)
Comment (arielvalentin) [reply instrumentation/pg/test/opentelemetry/instrumentation/pg/patches/connection_test.rb:12]: NVM Looks like they are... 😕

**Review by kaylareopelle — COMMENTED** [review]
This looks good to me once `require 'active_record'` is removed from the tests.

**Comment by Capncavedan** [comment]
Fantastic. I’ll incorporate feedback early next week (I’m out this week) and update the PR. -DanOn Jul 25, 2023, at 14:12, kaylareopelle ***@***.***> wrote:﻿
@kaylareopelle commented on this pull request.

This looks good to me once require 'active_record' is removed from the tests.

**Review by plantfansam — APPROVED** [review]
I think this makes sense! Thank you for the tests ❤️

**Review by arielvalentin — COMMENTED** [review]
Forgot to leave my review. I did some exploratory testing using the `example/pg.rb` script and things look good!

Thanks for your contribution. I will get a release out shortly.

**Comment by Capncavedan** [comment]
Thanks @arielvalentin !

**Pull request merged**